compiler/src/iree/compiler/Codegen/Common/FoldTensorSubsetIntoVectorTransferOps.cpp

```diff
@@ -204,11 +205,99 @@ class FoldInsertSliceIntoTransferWrite final
   }
 };
 
+/// Fold tensor.extract_slice into vector.transfer_write if
+///   1. The tensor.extract_slice op has only one use.
+///   2. All the offests of the tensor.extract_slice op are zeros.
+///   3. The vector.transfer_write op does not have masks.
+///   4. The vector.transfer_write op writes to a tensor.empty op.
+///
+/// E.g.:
+///
+/// ```
+/// %0 = vector.transfer_write %v, %t[%a, %b, %c]
+///   {in_bounds = [true, true, true]}
+///   : vector<1x64x128xf16>, tensor<1x64x128xf16>
+/// %extracted_slice = tensor.extract_slice %0[0, 0, 0] [1, %3, 128] [1, 1, 1]
+///   : tensor<1x64x128xf16> to tensor<1x?x128xf16>
+/// ```
+/// is rewritten to:
+/// ```
+/// %1 = vector.transfer_write %v, %t2[%a, %b, %c]
+///   {in_bounds = [true, false, true]}
+///   : vector<4x5xf32>, tensor<?x?xf32>
+/// ```
+class FoldExtractSliceIntoXferWrite final
+    : public OpRewritePattern<tensor::ExtractSliceOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractSliceOp extractSliceOp,
+                                PatternRewriter &rewriter) const override {
+    if (extractSliceOp.getDroppedDims().any()) {
+      return failure();
+    }
+
+    auto result = dyn_cast<OpResult>(extractSliceOp.getSource());
+    if (!result) {
+      return failure();
+    }
+    if (!result.hasOneUse()) {
+      return failure();
+    }
+    if (!llvm::all_of(extractSliceOp.getMixedOffsets(), isZeroIndex)) {
+      return failure();
+    }
+
+    auto xferOp =
+        extractSliceOp.getSource().getDefiningOp<vector::TransferWriteOp>();
+    if (!xferOp) {
+      return failure();
+    }
+    if (!xferOp.getSource().getDefiningOp<tensor::EmptyOp>()) {
+      return failure();
+    }
+    if (xferOp.getMask()) {
+      return failure();
+    }
+    if (!xferOp.getInBounds()) {
+      return failure();
+    }
```

Comment:
The default for `InBounds` is all false right? The below logic looks to conservatively handle that fine to me, is this check needed?

Reply:
Right, what I meant was that when `in_bounds` is empty, that is equivalent to `in_bound = [false, ..., false]`, so there is no need for this check because you're already handling the all false case.

Reply:
I see, makes sense to me now. Thank you!